### DISCUSSION TITLE
How to execute view-unrelated code before Next.js hydration in client-side?

### DISCUSSION THREAD

**Discussion by Brandon-Ln** [vercel/next.js#59189] [Help]
## TL, DR

I would like to know if this is feasible. If it is, could the warm-hearted community recommend some good practices? If this use case isn't feasible, maybe explanations or considerations regarding the design limitations would be nice.

## Details

I am new to using Next.js and currently attempting to migrate some traditional CSR applications to Next.js with SSG. I am using the pages router, but have encountered some usage issues during the migration process.

In my existing codebase, there are some code, such as handling polyfills with dynamic `import()` or fetching certain business configurations. Broadly speaking, these can be understood as side effects unrelated to the view layer, and this code will run before React performs client-side rendering.

```javascript
function polyfill() {
  if(typeof IntersectionObserver === "undefined"){
    return import("intersection-observer-polyfill")
  }
}

async function updateBizConfigAsync() {
  const config = inDom ? await (readNativeConfig() : Promise.resolve(undefined))

  if(config){
    updateConfig(config)
  }
}
// ...
```

and this code will run before React performs client-side rendering:

```javascript
// pages/entry.js

/**
 * You can think the `executeTasks` as a wrapper function of Promise.all().
 */
executeTasks(
  polyfill,
  updateBizConfigAsync,
  ...extraTasks
).then(() => {
  createRoot(document.getElementById("root")!).render(<App />)
}).catch(e => {
  /**
   * error handling
   */
})
```

However, this paradigm seems challenging to implement after migrating to Next.js. The reason is that Next.js appears to put a constraint that the expression of logic should be placed within the circumstances of components. since it only allows a default export(a component) of a page's entry. As a user, it seems that I have no way to touch lifecycle hooks such as `beforeClientSideHydration()`.

In my use case, I have tried some personal solutions, but it seems that they all have certain problems and are not ideal.

1. **Moving these side effect code outside of the component**：

```javascript

executeTasks()

export default function myApp() {
  // ...
}
```

Simply moving them outside of the component may not guarantee the execution order; the `executeTasks()` function is likely to be resolved after the client-side hydration of `myApp`, potentially causing incorrect business logic in the `useEffect` of children within `myApp`.

2. **use top-level `await`**：

```javascript
await executeTasks()

export default function myApp() {
  // ...
}
```

Now the execution order of the modules is guaranteed, and it seems to work correctly in my use case. Fortunately, Next.js itself also provides support for top-level `await`. However, I am unsure whether this is a misuse regarding ES features or a potential footgun. I am also curious whether it was impossible to achieve before ES2022.

3. **use root level `useEffect()` and `Context`**：

```javascript
export default function myApp() {
  // `useExternalExecuteTasksEffectState` wraps the resolved status with `useState` and `useEffect`
  const isResolved = useExternalExecuteTasksEffectState();

  return (
    <PrologueContext.Provider value={isResolved}>
      <RealApp />
    </PrologueContext.Provider>
  );
}
```

This solution seems intuitively straightforward to think of, but its downside is that even if `executeTasks` is just a set of side effects strongly related to business logic but UI-unrelated, such as updating singletons, conditionally polyfilling, or fetching network configurations based on the device, these tasks are now tightly coupled with React components. What I find more tricky is that all child components in `myApp` now need `useContext` to be aware of state changes, which seems to introduce a significant level of invasiveness to the business code.

In addition, I have searched through some discussions and issues, but have not found a satisfactory solution. Therefore, I think I might have missed sth, and maybe the community could provide some assistance. **It would be much appreciated if someone could address my confusion**.

**Comment by icyJoseph** [comment]
I guess in Pages Router, one could've had a `App.getInitialProps` in async mode, and run that whenever the context shows that we are client side, checking context.req or context.res, and your `executeTasks` function can make sure internally to run only once. 

You could also call the executeTasks function high up in the file, and await it it within getInitialProps.

Generally though, you could have defined an API that took in callbacks, and executed them whenever your business logic setup is ready, it is a bit more complex, but more robust too, and framework agnostic.

**Reply by Brandon-Ln** [reply]
Thx so much for your response! I did some research about `getInitialProps`, but I found the current docs said:

>**For the initial page load, getInitialProps will run on the server only**. getInitialProps will then also run on the client when navigating to a different route with the [next/link](https://nextjs.org/docs/pages/api-reference/components/link) component or by using [next/router](https://nextjs.org/docs/pages/api-reference/functions/use-router).

So the `getInitialProps` may probably not fit my use case? As I mentioned before, I use Next.js in SSG mode. So the behavior I want to achieve is to make the `executeTasks` execute when the first screen loaded on the client side. The `executeTasks` can be executed in CI in my use case, it probably has no actual meaning but is ok, since it runs just for isomorphism and does not affect the first-screen rendered result. Please feel free to correct me if I understand sth wrong about `getInitialProps`.

For your constructive advice:

>Generally though, you could have defined an API that took in callbacks, and executed them whenever your business logic setup is ready, it is a bit more complex, but more robust too, and framework agnostic.

Could you just explain it in more detail or show some snippets? The trouble I face is that when Next.js is executed on the client, I cannot guarantee that my setup logic is completed before Next.js App starts hydration. For my scenario, the setup logic will not affect the first screen hydration results of the Next.js, but it will cause, for example, some outdated business config to be read in `useEffect`(becaz the setup is not finished yet), or the dynamic polyfill/monitor script not to work correctly.

**Comment by MartinJaskulla** [comment]
It would be nice if `next.config.mjs` had something like Stencil's [`globalScript`](https://stenciljs.com/docs/config#globalscript) config option, which allows you to run a JavaScript file/function once before any other client side code runs.

I would like a solution that works the same way in both pages and app router.
